File: APIBanHangReactNative/models/model_voucher.js
const mongoose = require("mongoose");

const voucherSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true,
        trim: true,
    },
    code: {
        type: String,
        required: false, // Optional
        unique: true,
        uppercase: true,
        trim: true,
    },
    discountType: {
        type: String,
        enum: ["percentage", "fixed"], // Allowed values
        required: true,
        default: "percentage",
    },
    discountValue: {
        type: Number,
        required: true,
        validate: {
            validator: function (value) {
                return value > 0; // Ensure positive discount value
            },
            message: "Discount value must be greater than 0.",
        },
    },
    condition: {
        type: String,
        required: false, // Optional
        default: "No conditions",
    },
    startDate: {
        type: Date,
        required: true,
    },
    endDate: {
        type: Date,
        required: true,
        validate: {
            validator: function (value) {
                return value > this.startDate; // End date must be after start date
            },
            message: "End date must be after the start date.",
        },
    },
    quantity: {
        type: Number,
        required: true,
        min: [1, "Quantity must be at least 1"],
    },
    createdAt: {
        type: Date,
        default: Date.now, // Automatically sets creation date
    },
});

const Voucher = mongoose.model("Voucher", voucherSchema);

module.exports = Voucher;

File: app_banhang2/lib/services/models/model_voucher.dart
class ModelVoucher {
  final String id;
  final String name;
  final String code;
  final String discountType;
  final double discountValue;
  final String? condition;
  final DateTime startDate;
  final DateTime endDate;
  final int quantity;

  ModelVoucher({
    required this.id,
    required this.name,
    required this.code,
    required this.discountType,
    required this.discountValue,
    this.condition,
    required this.startDate,
    required this.endDate,
    required this.quantity,
  });

  factory ModelVoucher.fromJson(Map<String, dynamic> json) {
    return ModelVoucher(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      code: json['code'] ?? '',
      discountType: json['discountType'] ?? 'percentage',
      discountValue: double.tryParse(json['discountValue'].toString()) ?? 0.0,
      condition: json['condition'],
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      quantity: json['quantity'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
        '_id': id,
        'name': name,
        'code': code,
        'discountType': discountType,
        'discountValue': discountValue,
        'condition': condition,
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
        'quantity': quantity,
      };

  bool isValid() {
    final now = DateTime.now();
    return now.isAfter(startDate) && 
           now.isBefore(endDate) && 
           quantity > 0;
  }
}